apis/v1alpha1/nginxproxy_types.go

```diff
@@ -0,0 +1,91 @@
+// to configure global settings for all Gateways defined from the GatewayClass.
+type NginxProxy struct { //nolint:govet // standard field alignment, don't change it
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of the NginxProxy.
+	Spec NginxProxySpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// NginxProxyList contains a list of NginxProxies.
+type NginxProxyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []NginxProxy `json:"items"`
+}
+
+// NginxProxySpec defines the desired state of the NginxProxy.
+type NginxProxySpec struct {
+	// Telemetry specifies the OpenTelemetry configuration.
+	//
+	// +optional
+	Telemetry *Telemetry `json:"telemetry,omitempty"`
+}
+
+// Telemetry specifies the OpenTelemetry configuration.
+type Telemetry struct {
+	// Exporter specifies OpenTelemetry export parameters.
+	//
+	// +optional
+	Exporter *TelemetryExporter `json:"exporter,omitempty"`
+
+	// ServiceName is the "service.name" attribute of the OpenTelemetry resource.
+	// Default is 'ngf:<gateway-namespace>:<gateway-name>'. If a value is provided by the user,
+	// then the default becomes a prefix to that value.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxLength=127
+	ServiceName *string `json:"serviceName,omitempty"`
+
+	// SpanAttributes are custom key/value attributes that are added to each span.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxProperties=64
+	SpanAttributes map[string]AttributeValue `json:"spanAttributes,omitempty"`
+}
+
+// TelemetryExporter specifies OpenTelemetry export parameters.
+type TelemetryExporter struct {
+	// Interval is the maximum interval between two exports.
+	// Default: https://nginx.org/en/docs/ngx_otel_module.html#otel_exporter
+	//
+	// +optional
+	Interval *Duration `json:"interval,omitempty"`
+
+	// BatchSize is the maximum number of spans to be sent in one batch per worker.
+	// Default: https://nginx.org/en/docs/ngx_otel_module.html#otel_exporter
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	BatchSize *int32 `json:"batchSize,omitempty"`
+
+	// BatchCount is the number of pending batches per worker, spans exceeding the limit are dropped.
+	// Default: https://nginx.org/en/docs/ngx_otel_module.html#otel_exporter
+	//
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	BatchCount *int32 `json:"batchCount,omitempty"`
+
+	// Endpoint is the address of OTLP/gRPC endpoint that will accept telemetry data.
+	//
+	//nolint:lll
+	// +kubebuilder:validation:Pattern=`^(?:http?:\/\/)?[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?)*(?::\d{1,5})?$`
+	Endpoint string `json:"endpoint"`
+}
+
+// AttributeValue is a value paired with a key and attached to a tracing span.
+//
+// +kubebuilder:validation:MaxLength=255
+type AttributeValue string
```

Comment:
what about validating the value of attribute?

Reply:
What do you propose? Disallowing certain characters?

Reply:
I could add regex to disallow special characters

Reply:
in general, if we put the value in "..." in NGINX config, we can use the following approach -- https://github.com/nginxinc/nginx-gateway-fabric/blob/f3e9b9c1c478ace609126ccd9bd7a9bebbe2472b/internal/mode/static/nginx/config/validation/common.go#L33. , providing that variables are not supported by that directive.

If variables are supported, we can prevent interpolation of them https://github.com/nginxinc/nginx-gateway-fabric/blob/f3e9b9c1c478ace609126ccd9bd7a9bebbe2472b/internal/mode/static/nginx/config/validation/common.go#L52 (so that NGINX will not have a chance to fail to reload interpolating an invalid variable. Example. "value$somevar", where $somevar variable doesn't exist in NGINX.

Reply:
I guess it comes down to whether we want to restrict special characters in the CRD, or perform this validation in NGF using the function you mention.

Reply:
For now I can just keep the same restriction as with service name (alphanumeric characters with hyphens and underscores), and if we get feedback in the future to expand that, then we can loosen restrictions.